### ISSUE TITLE
Possible race condition for exceptions from spawner start before progress page redirect

### ISSUE THREAD

**Issue by behrmann** [jupyterhub/jupyterhub#4244]
### Bug description

I'm currently writing a spawner and implemented the [needed methods](https://jupyterhub.readthedocs.io/en/stable/reference/spawners.html#exception-handling) to report issues on the spawner progress screen, but I've been seeing a lot of non-descript "500 Internal Server error" lately.

Putting in a `time.sleep(1)`  at the beginning of the spawner's start method, I can see in the log that without the additional sleep the exception happens before the line `302 GET /hub/spawn/<username> -> /hub/spawn-pending/<username>` entry is written. 

Since the printing of the error message seems to be handled by the progress page, a race condition between an exception in the spawner's start method and the redirect to the progress page looks like the culprit to me.

#### Expected behaviour

Error messages from the spawner are shown to the user.

#### Actual behaviour

Error messages from the spawner are not shown to the user, only the "500 Internal Server error" message is shown.

### How to reproduce

I'm experiencing this with [my spawner](https://github.com/behrmann/systemdspawner/tree/staticspawner) with the checks at the beginning of the start method. A `time.sleep(1)` right at the start of the method works around this, but that's not a solution.

I guess any spawner that throws an exception right at the beginning of the start method could be used to reproduce this.

### Your personal set up

A current Debian bullseye with systemd v251 installed from backports and polkit backported from experimental, [my custom spawner](https://github.com/behrmann/systemdspawner/tree/staticspawner) and jupyterhub and all of its dependencies installed in a virtual environment with pip, using the system python (3.9).

<details><summary>Full environment</summary>
```
alembic==1.8.1
anyio==3.6.2
argon2-cffi==21.3.0
argon2-cffi-bindings==21.2.0
asteval==0.9.28
asttokens==2.2.0
async-generator==1.10
attrs==22.1.0
Babel==2.11.0
backcall==0.2.0
beautifulsoup4==4.11.1
black==22.10.0
bleach==5.0.1
certifi==2022.9.24
certipy==0.1.3
cffi==1.15.1
charset-normalizer==2.1.1
click==8.1.3
colorama==0.4.6
contourpy==1.0.6
cryptography==38.0.4
cycler==0.11.0
Cython==0.29.32
debugpy==1.6.4
decorator==5.1.1
defusedxml==0.7.1
entrypoints==0.4
executing==1.2.0
fastjsonschema==2.16.2
fonttools==4.38.0
future==0.18.2
gitdb==4.0.10
GitPython==3.1.29
greenlet==2.0.1
idna==3.4
importlib-metadata==5.1.0
ipykernel==6.17.1
ipympl==0.9.2
ipython==8.7.0
ipython-genutils==0.2.0
ipywidgets==8.0.2
isort==5.10.1
jedi==0.18.2
Jinja2==3.1.2
joblib==1.2.0
json5==0.9.10
jsonschema==4.17.3
jupyter-client==7.4.7
jupyter-core==5.1.0
jupyter-server==1.23.3
jupyter-server-mathjax==0.2.6
jupyter-telemetry==0.1.0
jupyterhub==3.0.0
jupyterhub-ldapauthenticator==1.3.2
-e git+https://github.com/behrmann/systemdspawner@f9db5b43a7983c2476aa0dc375af69f754e80ca3#egg=jupyterhub_systemdspawner
jupyterlab==3.5.0
jupyterlab-code-formatter==1.5.3
jupyterlab-git==0.40.1
jupyterlab-pygments==0.2.2
jupyterlab-server==2.16.3
jupyterlab-widgets==3.0.3
kiwisolver==1.4.4
ldap3==2.9.1
lmfit==1.1.0
Mako==1.2.4
MarkupSafe==2.1.1
matplotlib==3.6.2
matplotlib-inline==0.1.6
mistune==2.0.4
mpmath==1.2.1
mypy-extensions==0.4.3
nbclassic==0.4.8
nbclient==0.7.2
nbconvert==7.2.5
nbdime==3.1.1
nbformat==5.7.0
nest-asyncio==1.5.6
notebook==6.5.2
notebook-shim==0.2.2
numexpr==2.8.4
numpy==1.23.5
oauthlib==3.2.2
packaging==21.3
pamela==1.0.0
pandas==1.5.2
pandocfilters==1.5.0
parso==0.8.3
pathspec==0.10.2
pexpect==4.8.0
pickleshare==0.7.5
Pillow==9.3.0
platformdirs==2.5.4
prometheus-client==0.15.0
prompt-toolkit==3.0.33
psutil==5.9.4
psycopg2==2.9.5
ptyprocess==0.7.0
pure-eval==0.2.2
pyasn1==0.4.8
pycparser==2.21
Pygments==2.13.0
pyOpenSSL==22.1.0
pyparsing==3.0.9
pyrsistent==0.19.2
python-dateutil==2.8.2
python-json-logger==2.0.4
pytz==2022.6
pyzmq==24.0.1
qutip==4.7.0
requests==2.28.1
ruamel.yaml==0.17.21
ruamel.yaml.clib==0.2.7
scikit-learn==1.1.3
scipy==1.9.3
Send2Trash==1.8.0
six==1.16.0
smmap==5.0.0
sniffio==1.3.0
soupsieve==2.3.2.post1
SQLAlchemy==1.4.44
stack-data==0.6.2
sympy==1.11.1
terminado==0.17.0
threadpoolctl==3.1.0
tinycss2==1.2.1
tomli==2.0.1
tornado==6.2
traitlets==5.6.0
typing-extensions==4.4.0
uncertainties==3.1.7
urllib3==1.26.13
wcwidth==0.2.5
webencodings==0.5.1
websocket-client==1.4.2
widgetsnbextension==4.0.3
zipp==3.11.0
```
</details>

**Comment by minrk** [comment]
Thanks for the report!

There's definitely a race in that visiting spawn-pending for a server that's already failed and a server that hasn't yet failed will produce slightly different views of the failure (one will have a sequence of progress events until the failure, while the latter only shows a 'last spawn failed' page. Importantly, the 'last spawn failed' page does _not_ include the error message, while the progress page does. These are meant for requests separated by a significant amount of time, but if failure happens very quickly (i.e. within one second), a regular spawn sequence looks like an implicit spawn attempt for a server that failed a while ago.

So we have some possible ways to solve this:

1. save more about the failed spawn (could be a bit tricky, since progress events aren't persisted on the server side), so the failed page looks identical whether the launch has already failed or not
2. render the 'spawner already failed' situation a bit differently, so it looks more like the spawn-pending page, but with only the failure event present. This probably offers the best balance, but has the downside that attempting to visit a server that failed to launch long ago _looks_ like it failed just now.
3. (smallest change) include the same error message that the progress event shows in the "last spawn failed" page. This solves the 'informative error message' is gone problem, but leaves the 'fast failure looks different to slow failure' race problem.
4. special-case the _first_ request for the spawn-pending page after a failure (solve the race only, preserve remaining behavior) so that it always looks like a pending launch

**Comment by behrmann** [comment]
Sorry for the late reply, a different project kept me busy this week so far.

Let me rephrase what your saying, so that I know whether I get it. What I'm seeing when the spawner fails too quickly, is the `not_running.html`, because my "last spawn" (what I'm thinking of as the current spawn) failed. 

I'm not sure I understand the difference between the third and the first solution. Since showing the most recent error message would require persisting the progress events in some way (or at least the last failure), do you mean to add the on event stuff from the `spawn_pending.html` to `not_running.html`?

**Comment by minrk** [comment]
The difference is persisting a single error message (we already do this, so the only change is _showing_ it) vs persisting a potentially long _stream_ of events that is currently not recorded anywhere, nor used anywhere except in the event of a failure.

**Comment by behrmann** [comment]
I think showing a single error should be sufficient, considering that for now it's none. :)
